lib/course_api.rb

```diff
@@ -0,0 +1,143 @@
+            course_title = course[:title]
+            cids = []
+            course.xpath(".//class").each do |cl|
+                cids << remove_class_id_prefix(cl[:id])
+              end
+            course.xpath(".//class").each do |class_obj|
+                request_class_id = class_obj[:id]
+                class_id = remove_class_id_prefix(request_class_id)
+                # Get all the sections for this course which have instructors
+                #puts "Getting sections for " + request_class_id.to_s
+                
+                # Get sections using the Course API, look only for sections
+                # that have instructors lister
+                # After a certain amount of calls, wait a second
+                counter += 1
+                #puts counter
+                if(counter == 10000)
+                    puts "reset counter"
+                    sleep(1)
+                    counter = 0
+                end 
+                sections = request_course_API(request_class_id)
+                sections.each do |section|
+                    section_id = section[:sid]
+                    instructors = section[:instructors]
+                    course_hash = {"title": course_title, "term": term_display, "cid": class_id, "cids": cids, 
+                    "sid":section_id, "instructors": instructors}
+                    courses << course_hash
+                end
+            end 
+        end
+        # Array of course hash objects
+        courses
+    end
+
+    def request_course_API(request_class_id)
+        # Put in some time between calls to help with timeouts
+        cert_file = Settings.certs_path + "sul-harvester.cert"
+        key_file = Settings.certs_path + "sul-harvester.key"
+        sections = []
+        errors = []
+        client_cert = OpenSSL::X509::Certificate.new File.read(cert_file)
+        client_key = OpenSSL::PKey.read File.read(key_file)
+        connection = Faraday::Connection.new 'https://registry.stanford.edu', :ssl => { :client_cert => client_cert, :client_key => client_key }
+        spec_course = '/doc/courseclass/' + request_class_id
+        begin
+            response = connection.get spec_course
+        # Timeout error
+        rescue Faraday::TimeoutError
+            errors << "timeout error occurred for first attempt for " + request_class_id.to_s
+            # Retry
+            begin 
+                response = connection.get spec_course
+            rescue Faraday::Error
+                errors << "retry failed for " + request_class_id.to_s
+            end
+        # Connection failed error
+        rescue Faraday::ConnectionFailed
+            errors << "Connection failed occurred for first attempt for " + request_class_id.to_s
+        rescue Faraday::ServerError
+            errors << "Server Error for " + request_class_id.to_s
+        end
+        if response.status == 200
+            course_response = Nokogiri::XML(response.body)
```

Comment:
Can this XML parsing be extracted to a separate method or class? That would make the code easier to reason about.